Sources/Shared/LocalAuthentication/BiometricsAuthenticationService.swift

```diff
@@ -0,0 +1,142 @@
+import Foundation
+import LocalAuthentication
+import UIKit
+
+public protocol BiometricsAuthenticationServiceProtocol {
+    func checkForBiometrics(controller: UIViewController)
+    func protectAppIfNeeded(controller: UIViewController, completion: (() -> Void)?)
+}
+
+class BiometricsAuthenticationService: BiometricsAuthenticationServiceProtocol {
+    private var context: LAContext?
+
+    private var overlayViewController: BiometricsAuthenticationViewController?
+
+    private var shouldAllowMacToAuthenticate = true
+
+    init() {
+        #if targetEnvironment(macCatalyst)
+        if #available(iOS 13.0, *) {
+            NotificationCenter.default.addObserver(
+                self,
+                selector: #selector(allowReauthenticate),
+                name: UIScene.didEnterBackgroundNotification,
+                object: nil
+            )
+        } else {
+            NotificationCenter.default.addObserver(
+                self,
+                selector: #selector(allowReauthenticate),
+                name: UIApplication.didEnterBackgroundNotification,
+                object: nil
+            )
+        }
+        #endif
+    }
+
+    /// Adds the protection overlay and tries to authenticate
+    func checkForBiometrics(controller: UIViewController) {
+        guard context == nil else { return }
+        protectAppIfNeeded(controller: controller) { [weak self] in
+            if Current.settingsStore.biometricsRequired {
+                #if targetEnvironment(macCatalyst)
+                guard self?.shouldAllowMacToAuthenticate ?? false else { return }
+                #endif
+                self?.authenticate()
+            }
+        }
+    }
+
+    /// Adds the protection overlay
+    func protectAppIfNeeded(controller: UIViewController, completion: (() -> Void)?) {
+        if Current.settingsStore.biometricsRequired {
+            addBiometricOverlayProtection(controller: controller) {
+                completion?()
+            }
+        } else {
+            removeProtectionOverlay()
+        }
+    }
+
+    #if targetEnvironment(macCatalyst)
+    /// This avoids macOS going on loop between enter background and enter foreground while trying to authenticate
+    @objc private func allowReauthenticate() {
+        shouldAllowMacToAuthenticate = true
+    }
+    #endif
+
+    private func authenticate() {
+        var error: NSError?
+        context?.invalidate()
+        context = LAContext()
+        if context?.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) ?? false {
```

Comment:
should we allow watch too?

Reply:
I've never seen this happening inside apps, only unlocking iPhone itself, can you point me out to the policy?

Reply:
`kLAPolicyDeviceOwnerAuthenticationWithBiometricsOrWatch`

Reply:
This is available only for Mac, which this PR won't support yet.